tensorflow_probability/python/experimental/mcmc/particle_filter.py

```diff
@@ -435,13 +472,410 @@ def seeded_one_step(seed_state_results, _):
+              rej_kernel.one_step(
+                  rej_inner_weighted_particles, rej_filter_results, seed=seed
+              )
+
+          rej_parameters_weights += rej_inner_weighted_particles.log_weights
+
+          rej_params_log_weights = \
+              rej_params_log_weights + \
+              rej_filter_results.incremental_log_marginal_likelihood
+          return i + 1, \
+                 rej_inner_weighted_particles, \
+                 rej_filter_results, \
+                 rej_parameters_weights, \
+                 rej_params_log_weights
+
+        _, \
+        rej_inner_weighted_particles, \
+        rej_filter_results, \
+        rej_inner_particles_weights, \
+        rej_params_log_weights = tf.while_loop(
+            condition,
+            body,
+            loop_vars=[0,
+                       rej_inner_weighted_particles,
+                       rej_filter_results,
+                       rej_inner_particles_weights,
+                       rej_params_log_weights
+                       ]
+        )
+
+        log_a = rej_filter_results.accumulated_log_marginal_likelihood - \
+                filter_results.accumulated_log_marginal_likelihood + \
+                parameter_proposal_kernel(
+                    proposed_parameters).log_prob(outside_parameters) - \
+                parameter_proposal_kernel(
+                    outside_parameters).log_prob(proposed_parameters)
+
+        acceptance_probs = tf.minimum(1., tf.exp(log_a))
+
+        random_numbers = uniform.Uniform(0., 1.).sample(num_outer_particles,
+                                                        seed=seed)
+
+        # Determine if the proposed particle should be accepted or reject
+        accept = random_numbers > acceptance_probs
+
+        # Update the chosen particles and filter restults
+        # based on the acceptance step
+        outside_parameters = tf.where(accept,
+                                      outside_parameters,
+                                      proposed_parameters)
+        updated_log_weights = tf.where(accept,
+                                       updated_log_weights,
+                                       rej_params_log_weights)
+
+        inner_weighted_particles_particles = mcmc_util.choose(
+            accept,
+            inner_weighted_particles.particles,
+            rej_inner_weighted_particles.particles
+        )
+        inner_weighted_particles_log_weights = mcmc_util.choose(
+            accept,
+            inner_weighted_particles.log_weights,
+            rej_inner_weighted_particles.log_weights
+        )
+
+        inner_weighted_particles = smc_kernel.WeightedParticles(
+            particles=inner_weighted_particles_particles,
+            log_weights=inner_weighted_particles_log_weights,
+            extra=inner_weighted_particles.extra
+        )
+
+        filter_results = tf.nest.map_structure(
+            lambda a, b: where_fn(accept, a, b,
```

Comment:
We should be able to use `mcmc_util.choose` instead of this custom `where_fn`.

I think the errors you were seeing (which led to defining `where_fn`) were because the `steps` and `seed` fields in `SequentialMonteCarloResults` do not have the same batch shape as the other fields.  It should work to do something like:
```
(parent_indices,
 incremental_log_marginal_likelihood,
 accumulated_log_marginal_likelihood) = mcmc_util.choose(..., ...old fields..., new fields...)
```

I think it is not currently possible to set the seed correctly here, so taking the seed from either `filter_results` or `rej_filter_results` is okay.  (In a later cleanup, we could consider adding batch shape to `seed` to fix this.)